src/Shadowdomize/DOMFunctions.ts

```diff
@@ -0,0 +1,226 @@
+        activeElement = activeElement.shadowRoot.activeElement;
+    }
+
+    return activeElement;
+}
+
+export function nodeContains(
+    node: Node | null | undefined,
+    otherNode: Node | null | undefined
+): boolean {
+    if (!node || !otherNode) {
+        return false;
+    }
+
+    let currentNode: HTMLElement | Node | null | undefined = otherNode;
+
+    while (currentNode) {
+        if (currentNode === node) {
+            return true;
+        }
+
+        if (
+            typeof (currentNode as HTMLSlotElement).assignedElements !==
+                "function" &&
+            (currentNode as HTMLElement).assignedSlot?.parentNode
+        ) {
+            // Element is slotted
+            currentNode = (currentNode as HTMLElement).assignedSlot?.parentNode;
+        } else if (currentNode.nodeType === document.DOCUMENT_FRAGMENT_NODE) {
+            // Element is in shadow root
+            currentNode = (currentNode as ShadowRoot).host;
+        } else {
+            currentNode = currentNode.parentNode;
+        }
+    }
+
+    return false;
+}
+
+export function getParentNode(
+    node: Node | null | undefined
+): ParentNode | null {
+    if (!node) {
+        return null;
+    }
+
+    if (
+        node.nodeType === Node.DOCUMENT_FRAGMENT_NODE &&
+        (node as ShadowRoot).host
+    ) {
+        return (node as ShadowRoot).host;
+    }
+
+    return node.parentNode;
+}
+
+export function getParentElement(
+    element: HTMLElement | null | undefined
+): HTMLElement | null {
+    for (
+        let parentNode = getParentNode(element);
+        parentNode;
+        parentNode = getParentNode(parentNode)
+    ) {
+        if (parentNode.nodeType === Node.ELEMENT_NODE) {
+            return parentNode as HTMLElement;
+        }
+    }
+
+    return null;
+}
+
+export function getFirstChild(node: Node | null | undefined): ChildNode | null {
+    if (!node) {
+        return null;
+    }
+
+    if ((node as Element).shadowRoot) {
+        const child = getFirstChild((node as Element).shadowRoot);
+
+        if (child) {
+            return child;
+        }
+
+        // If the attached shadowRoot has no children, just try ordinary children,
+        // that might come after.
+    }
+
+    return node.firstChild;
+}
+
+export function getLastChild(node: Node | null | undefined): ChildNode | null {
+    if (!node) {
+        return null;
+    }
+
+    if (!node.lastChild && (node as Element).shadowRoot) {
+        return getLastChild((node as Element).shadowRoot);
+    }
+
+    return node.lastChild;
+}
+
+export function getNextSibling(
+    node: Node | null | undefined
+): ChildNode | null {
+    return node?.nextSibling || null;
+}
+
+export function getPreviousSibling(
+    node: Node | null | undefined
+): ChildNode | null {
+    if (!node) {
+        return null;
+    }
+
+    let sibling = node.previousSibling;
+
+    if (!sibling && node.parentElement?.shadowRoot) {
```

Comment:
I wonder if we need to handle this case? Nothing you append to the host of a shadow root seems to render anything https://codesandbox.io/p/sandbox/vanilla-js-forked-6964cz?file=%2Fsrc%2Findex.js%3A20%2C1

Reply:
Yes. I have added it before the realization that browser ignores all children of the shadow host, other than attached shadow. We can remove this once these functions are moved to the dedicated package and covered with dedicated tests.